### ISSUE TITLE
System.Windows.Automation.InvokePattern.Invoke blocks on a menu item.

### ISSUE THREAD

**Issue by lovettchris** [dotnet/winforms#10244]
### .NET version

7.0.403

### Did it work in .NET Framework?

Yes

### Did it work in any of the earlier releases of .NET Core or .NET 5+?

Don't know, it works in .NET Framework 4.6.2, but not in latest version of .NET Framework 4.8

### Issue description

Run the unit test in the following solution. Notice that the unit test hangs on line 80 of UnitTest1.cs which is the InvokePattern.Invoke call on the FIle/Open... menu item.  This makes it impossible to test any menu item using System.Automation?  

### Steps to reproduce

1. on Windows 11 22H2 build 22621.2361
2. Load this solution into VS 2022
3. Run Test/Debug all tests
4. wait for windows open file dialog to appear
5. Notice it never disappears
6. hit break in the debugger and you will see it is blocked on line 80 of UnitTest1.cs.

[WinFormsApp1.zip](https://github.com/dotnet/winforms/files/13289996/WinFormsApp1.zip)

**Comment by elachlan** [comment]
@Olina-Zhang can your team please test this to confirm?

**Comment by Amy-Li03** [comment]
@elachlan we can repro this issue on .NET 7.0, 8.0 and 9.0.
![Repro](https://github.com/dotnet/winforms/assets/62929087/1fa8a294-978a-429a-8acc-af3758db63c8)

**Comment by lovettchris** [comment]
[like]  Chris Lovett reacted to your message:
________________________________

**Comment by Epica3055** [comment]
I and @Amy-Li03 discussed this issue, and we think this is not a a11y issue.

**Comment by merriemcgaw** [comment]
@Epica3055 let's discuss in the next meeting.

**Comment by Tanya-Solyanik** [comment]
@Amy-Li03  - Could you please verify if this issue reproduces on 4.7.2 and 4.8.1 and if it depends on any AccessibilityImprovements.Level?

**Comment by lovettchris** [comment]
Attached is a .NET Framework 4.7.2 version of the above test app and the test passes (the invoke does **not** block)
[WindowsFormsApp1.zip](https://github.com/dotnet/winforms/files/13359676/WindowsFormsApp1.zip)

If you then change the project properties to target .NET Framework 4.8, the test fails as per my original bug report.

**Comment by Amy-Li03** [comment]
@Tanya-Solyanik , this issue reproduces on .NET Framework 4.8 & 4.8.1 and it doesn't repro on 4.7.2 and its previous versions. 
Also it doesn't depend on AccessibilityImprovements.Level, add following quirk to app.config file, issue still repro.
```
<runtime>
	<AppContextSwitchOverrides value="Switch.UseLegacyAccessibilityFeatures.3=false;Switch.UseLegacyAccessibilityFeatures.4=false;Switch.UseLegacyAccessibilityFeatures.5=false"/>
</runtime>
```

**Comment by LeafShi1** [comment]
**.netFramwork4.7 project:**
The function `RawInvokePattern_Invoke(hobj)` return 0

```
UIAutomationClientsideProviders.dll!MS.Internal.AutomationProxies.Accessible.DoDefaultAction() Line 527	C#
UIAutomationClientsideProviders.dll!MS.Internal.AutomationProxies.MsaaNativeProvider.CallDoDefaultAction() Line 820	C#
UIAutomationClientsideProviders.dll!MS.Internal.AutomationProxies.MsaaNativeProvider.System.Windows.Automation.Provider.IInvokeProvider.Invoke() Line 515	C#
[Native to Managed Transition]	
[Managed to Native Transition]	
 UIAutomationClient.dll!MS.Internal.Automation.UiaCoreApi.InvokePattern_Invoke(MS.Internal.Automation.SafePatternHandle hobj) Line 659	C#
```

**netFramwork4.8.1 project:**
The function `RawInvokePattern_Invoke(hobj)` return -2146233083
```
UIAutomationClient.dll!MS.Internal.Automation.UiaCoreApi.CheckError(int hr) Line 1017	C#
UIAutomationClient.dll!MS.Internal.Automation.UiaCoreApi.InvokePattern_Invoke(MS.Internal.Automation.SafePatternHandle hobj) Line 660	C#
UIAutomationClient.dll!System.Windows.Automation.InvokePattern.Invoke() Line 28	C#
TestProject1.dll!TestProject1.UnitTest1.Invoke(System.Windows.Automation.AutomationElement e) Line 99	C#
TestProject1.dll!TestProject1.UnitTest1.InvokeMenuItem(System.Windows.Automation.AutomationElement window, string menuItemName) Line 91	C#
TestProject1.dll!TestProject1.UnitTest1.TestMethod1() Line 35	C#
```

**Comment by Tanya-Solyanik** [comment]
@LeafShi1  - does 4.8.1 sample block as well?

**Comment by LeafShi1** [comment]
> @LeafShi1 - does 4.8.1 sample block as well?

Yes

**Comment by Tanya-Solyanik** [comment]
@LeafShi1  - what other controls does this reproduce for?

**Comment by LeafShi1** [comment]
> @LeafShi1 - what other controls does this reproduce for?

I tried this in a button, it work well. 
[WindowsFormsApp9.zip](https://github.com/dotnet/winforms/files/14902379/WindowsFormsApp9.zip)

**Comment by Tanya-Solyanik** [comment]
@LeafShi1 - do we have any unit tests that call the Invoke method on controls? If not, could you please add some?

**Comment by LeafShi1** [comment]
> @LeafShi1 - do we have any unit tests that call the Invoke method on controls? If not, could you please add some?

Many of controls contained IsPatternSupported test, but not call the Invoke method on controls? Did we need to complete all of them. Such as https://github.com/dotnet/winforms/blob/9313f05e84328c3176eb7d80497c9505cd8aaebf/src/System.Windows.Forms/tests/UnitTests/System/Windows/Forms/AccessibleObjects/HScrollBar.ScrollBarFirstLineButtonAccessibleObjectTests.cs#L128-L135

**Comment by Tanya-Solyanik** [comment]
Start with the MenuItems mentioned in this bug and buttons, we'll see if we need to add more later.

**Comment by LeafShi1** [comment]
> Start with the MenuItems mentioned in this bug and buttons, we'll see if we need to add more later.

In System.Windows.Forms.AccessibleObject, Button doesn't support InvokePattern, so only added a test for MenuItem.

**Comment by Tanya-Solyanik** [comment]
This is interesting, my test application was blocking on a button. I was doing a cross process access though. I have a driver application and a test application, and the driver is invoking buttons in the test app.

Let's test only for menuitems for now. Going forward we can start replacing code that send clicks to buttons(or other controls) with AccessibleObject.Invoke in tests that excersize control functionality.

**Comment by clovett** [comment]
Well after waiting for almost a year I've given up waiting.  I ended up implementing my own custom automation layer for my menu items so that I could invoke them asynchronously.  This gets my unit tests back up and running finally.
See this commit: https://github.com/microsoft/XmlNotepad/commit/484713b6bd527a27644109994c139f3dd007deb3
And specifically the [HiddenMenuItems ](https://github.com/microsoft/XmlNotepad/blob/master/src/XmlNotepad/HiddenMenuItems.cs) control that provides new custom AutomationObjects for my menu items with a feature that can invoke some of them asynchronously (specifically the ones that pop up a modal file open or save dialog).

**Comment by LeafShi1** [comment]
Thanks for sharing your test project and the follow-up commit.

After reviewing the provided code, we’ve concluded that the blocking behavior observed when invoking menu items that open modal dialogs (such as OpenFileDialog) is not caused by a bug in WinForms or UIAutomation. Instead, it stems from how the test case is structured.

Specifically, the test synchronously calls `InvokePattern.Invoke()` on a menu item that opens a modal dialog. According to the official documentation:

 **The behavior of this method depends entirely on the implementation of the UI Automation provider. If the call to Invoke causes a modal dialog to be displayed, the method might block.**   — [Microsoft Docs](https://learn.microsoft.com/en-us/dotnet/api/system.windows.automation.invokepattern.invoke?view=windowsdesktop-9.0)

This means that blocking is expected if the invoked action opens a modal dialog and the provider does not handle it asynchronously. In your test, the automation thread ends up waiting on a UI thread that’s blocked by the modal dialog — a classic deadlock scenario.

Your workaround using a custom automation layer with asynchronous invocation is a valid and effective solution for this case. Since the issue arises from the test architecture rather than a framework defect, we’ll go ahead and close this issue.

**Comment by lovettchris** [comment]
This is wrong, and the behavior has clearly changed recently which at best is a huge breaking change.  But worse, when it blocks like this there is no way for the client to automate the dialog itself.

**Comment by LeafShi1** [comment]
@merriemcgaw  @KlausLoeffelmann This issue started with .NET Framework 4.8. Do you have any suggestions?

**Comment by lovettchris** [comment]
Dig deeper into why it changed.
